File: apps/brunch-agent/test/reopened-why-retention-audit.py
#!/usr/bin/env python3
"""Read-only A5 observation audit and falsifiers. Never opens, copies or imports a DB/profile.

The primary oracle remains the built ChatAgent test. This independently checks its retained
outputs and rejects coordinated omissions across snapshots using pre-compaction canonical pins.
Accepts original JSON or losslessly gzipped evidence, without restoring anything into an app.
"""
import copy
import gzip
import hashlib
import json
from pathlib import Path
import sys


def require(condition, message):
    if not condition:
        raise AssertionError(message)


def text(message):
    return "".join(part["text"] for part in message["parts"] if part["type"] == "text")


def tools(snapshot):
    return [part for message in snapshot["messages"] for part in message["parts"] if part["type"] == "dynamic-tool"]


def tool(snapshot, call_id):
    found = [part for part in tools(snapshot) if part["toolCallId"] == call_id]
    require(len(found) == 1 and found[0]["state"] == "output-available", f"exact completed tool: {call_id}")
    return found[0]


snapshot_names = ["create-history", "fold-before", "fold-immediate-history", "fold-history", "reopen-before", "reopen-history"]
query_names = ["process-restarted-before-fold", "fold-after-compaction", "reopen-after-compaction"]
names = snapshot_names + query_names + ["seed", "fold-completion-pins", "fold-canonical-settlements", "reopen-canonical-settlements", "fold-contexts", "reopen-contexts", "fold-events"] + [f"{phase}-{kind}" for phase in ["create", "fold", "reopen"] for kind in ["process", "result"]]


def audit(data):
    seed = data["seed"]
    pids = [data[f"{phase}-process"]["pid"] for phase in ["create", "fold", "reopen"]]
    require(len(set(pids)) == 3, "distinct actual process IDs")
    for phase, pid in zip(["create", "fold", "reopen"], pids):
        require(data[f"{phase}-result"]["pid"] == pid, "process/result PID correlation")
        require(data[f"{phase}-process"]["dbPath"] == seed["dbPath"], "same original store path")
        require(data[f"{phase}-result"]["outcome"] == "pass", "actual phase pass")
    require(seed["pid"] == pids[0], "seed PID correlation")
    baseline = data["create-history"]
    source = [message for message in baseline["messages"] if message["id"] == seed["sourceId"]]
    require(len(source) == 1, "source exact identity/content")
    source = source[0]
    require(source["role"] == "user" and source["purpose"] == "user", "source authorized role/purpose")
    source_text = "TEST synthetic original testimony control: When final inspection starts, reserve one available crew until sign-off."
    require(text(source) == source_text, "source exact identity/content")
    protected = [part for part in tools(baseline) if part["toolName"] in ["update_workpiece", "addArc", "getLatestNetDefinition"]]
    require(len(protected) == 6, "three revisions, two reads, one mutation; no tool reissue")
    for number in [1, 2, 3]:
        revision = tool(baseline, f"retention-revision-{number}")
        pointer = revision["output"]
        require(pointer["revisionId"] == revision["toolCallId"] and pointer["ordinal"] == number, "revision identity/ordinal")
        require(hashlib.sha256(revision["input"]["markdown"].encode()).hexdigest() == pointer["sha256"], "revision content hash")
        require(pointer["evidenceValidated"] is True, "validated revision evidence")
        require(pointer["evidence"] == seed["governing"]["evidence"] and len(pointer["evidence"]) == 2, "overlapping carried evidence exact")
        if number > 1:
            require("evidence" not in revision["input"], "raw carried input not rewritten")
    original = tool(baseline, "retention-live-why")["output"]
    require(original["reconciliation"]["status"] == "live-observed", "creation actual live observation")
    for name in snapshot_names:
        snapshot = data[name]
        require([message for message in snapshot["messages"] if message["id"] == seed["sourceId"]] == [source], "source exact identity/content")
        require([part for part in tools(snapshot) if part["toolName"] in ["update_workpiece", "addArc", "getLatestNetDefinition"]] == protected, "protected tools exact; no reissue")
        for message in baseline["messages"]:
            require([entry for entry in snapshot["messages"] if entry["id"] == message["id"]] == [message], "baseline public messages exact")
    for name in query_names:
        query = data[name]
        require(query["read"]["currentWorkpiece"] == original["currentWorkpiece"], "actual current state exact")
        for label in ["why", "oldObservationWhy"]:
            answer = query[label]
            require(answer["governing"] == original["governing"], "governing revision/hash/passages/relations exact")
            require(answer["recordedChange"] == original["recordedChange"], "actual recorded effects exact")
            require(answer["reconciliation"]["status"] == "as-of", "restart is as-of, not fresh browser")
            require(answer["disposition"] == "partially-supported" and answer["untrusted"] is True, "honest partial untrusted standing")
        require(query["oldObservationWhy"]["reconciliation"]["observationScope"] == "as-of", "old ID cannot earn freshness")
        require(query["refusedObservationWhy"]["disposition"] == "refused", "unknown observation refuses")
        if name != "process-restarted-before-fold":
            require(query["read"]["earlierSourcesOmitted"] > 0 and all(item["id"] != seed["sourceId"] for item in query["read"]["sources"]), "source window limit explicit")
            phase = "fold" if name.startswith("fold") else "reopen"
            request = data[f"{phase}-contexts"][query["beforeRequestContextIndex"]]
            require(request["purpose"] == "agent", "actual query request context")
            serialized = json.dumps(request["context"]["messages"])
            require("A5 controlled lossy summary" in serialized, "real folded summary consumed")
            require(source_text not in serialized, "original true-user source entry absent from query context")
            require(not any(message.get("role") == "toolResult" and message.get("toolName") in ["brunch_workpiece", "brunch_why"] for message in request["context"]["messages"]), "prior workpiece/why results absent from query context")
            require(not any(call_id in serialized for call_id in query["priorQueryIds"]), "prior query IDs absent even with redacted source text")
            # Authoritative current revision remains injected by the product, including
            # passage and evidence pointers. This is not source-entry or answer retention.
    starts = [event for event in data["fold-events"] if event["type"] == "compaction_start"]
    compactions = [event for event in data["fold-events"] if event["type"] == "compaction"]
    require(len(starts) >= 2 and all(event["reason"] == "threshold" for event in starts), "real threshold compaction, never overflow substitute")
    require(len(compactions) >= 2 and all(not event["isError"] and event["messagesAfter"] < event["messagesBefore"] for event in compactions), "successful real context folding")
    pins = data["fold-completion-pins"]
    require(len(pins) >= 22, "nonempty independent completion pins")
    for pin in pins:
        records = pin["records"]
        start = [record for record in records if record["type"] == "assistant_message_started"]
        end = [record for record in records if record["type"] == "assistant_message_completed"]
        require(len(start) == len(end) == 1, "canonical completion exists exactly once")
        start, end = start[0], end[0]
        body = "".join(record["delta"] for record in records if record["type"] == "assistant_text_delta")
        expected = dict(id=start["messageId"], role="assistant", purpose="assistant", display="visible", submissionId=start["submissionId"], turnId=start["turnId"], parts=[dict(type="text", text=body, state="done")])
        require(pin["message"] == expected and end["messageId"] == start["messageId"] and end["stopReason"] == "stop", "pin matches independent canonical completion")
        require(pin["event"]["turnId"] == start["turnId"] and pin["event"]["submissionId"] == start["submissionId"], "completion event correlation")
        for name in ["fold-history", "reopen-before", "reopen-history"]:
            snapshot = data[name]
            require([message for message in snapshot["messages"] if message["id"] == expected["id"]] == [expected], "independently pinned completed response exact")
            require([item for item in snapshot["settlements"] if item["submissionId"] == expected["submissionId"]] == [dict(submissionId=expected["submissionId"], outcome="completed", answeredBySubmissionId=expected["submissionId"])], "independently pinned completed settlement exact")
        for phase in ["fold", "reopen"]:
            settlements = [item for item in data[f"{phase}-canonical-settlements"] if item["submissionId"] == expected["submissionId"]]
            require(len(settlements) == 1 and settlements[0]["outcome"] == "completed", "canonical settlement exact")
    return dict(pids=pids, sameOriginalStore=seed["dbPath"], completionPins=len(pins), thresholdCompactions=len(compactions), sourceId=seed["sourceId"], governingRevision=seed["governing"]["revisionId"])


def falsify(original):
    results = []
    for mode, expected in [
        ("source-omission", "source exact"), ("source-change", "source exact"),
        ("carried-relation-loss", "overlapping carried evidence"),
        ("completion-omission", "independently pinned completed response"),
        ("completion-duplicate", "independently pinned completed response"),
        ("settlement-omission", "independently pinned completed settlement"),
        ("same-pid", "distinct actual process"), ("false-live", "restart is as-of"),
        ("source-still-in-context", "original true-user source entry absent"),
        ("redacted-answer-still-in-context", "prior workpiece/why results absent"),
    ]:
        data = copy.deepcopy(original)
        source_id = data["seed"]["sourceId"]
        pin = data["fold-completion-pins"][0]["message"]
        for name in snapshot_names:
            snapshot = data[name]
            if mode == "source-omission":
                snapshot["messages"] = [message for message in snapshot["messages"] if message["id"] != source_id]
            if mode == "source-change":
                for message in snapshot["messages"]:
                    if message["id"] == source_id:
                        message["parts"] = [dict(type="text", text="Mutated source", state="done")]
            if mode == "carried-relation-loss":
                tool(snapshot, "retention-revision-2")["output"]["evidence"].pop()
            if mode == "completion-omission":
                snapshot["messages"] = [message for message in snapshot["messages"] if message["id"] != pin["id"]]
            if mode == "completion-duplicate" and any(message["id"] == pin["id"] for message in snapshot["messages"]):
                snapshot["messages"].append(copy.deepcopy(pin))
            if mode == "settlement-omission":
                snapshot["settlements"] = [item for item in snapshot["settlements"] if item["submissionId"] != pin["submissionId"]]
        if mode == "same-pid":
            data["reopen-process"]["pid"] = data["create-process"]["pid"]
        if mode == "false-live":
            data["reopen-after-compaction"]["why"]["reconciliation"]["status"] = "live-observed"
        if mode == "source-still-in-context":
            index = data["reopen-after-compaction"]["beforeRequestContextIndex"]
            data["reopen-contexts"][index]["context"]["messages"].append(dict(role="user", content=text(next(message for message in original["create-history"]["messages"] if message["id"] == source_id))))
        if mode == "redacted-answer-still-in-context":
            index = data["reopen-after-compaction"]["beforeRequestContextIndex"]
            answer = copy.deepcopy(data["reopen-after-compaction"]["why"])
            for passage in answer["governing"]["passages"]:
                for relation in passage["relations"]:
                    relation["sources"] = []
            data["reopen-contexts"][index]["context"]["messages"].append(dict(role="toolResult", toolName="brunch_why", toolCallId="retention-live-why", content=[dict(type="text", text=json.dumps(answer))]))
        try:
            audit(data)
        except AssertionError as error:
            require(expected in str(error), f"Wrong discriminator for {mode}: {error}")
            results.append(dict(mode=mode, rejected=True, reason=str(error)))
        else:
            raise AssertionError(f"Falsifier escaped: {mode}")
    return results


if __name__ == "__main__":
    directory = Path(sys.argv[1])
    def load(name):
        path = directory / f"{name}.json"
        return json.loads(path.read_bytes() if path.exists() else gzip.decompress(path.with_suffix(".json.gz").read_bytes()))
    data = {name: load(name) for name in names}
    print(json.dumps(dict(observationAudit=audit(data), falsifiers=falsify(data)), indent=2))

File: apps/brunch-agent/test/integration/reopened-why-retention-audit.ts
import assert from "node:assert/strict";
import { createHash } from "node:crypto";
import { existsSync, readFileSync } from "node:fs";
import { join } from "node:path";
import { gunzipSync } from "node:zlib";

const snapshotNames = [
  "create-history",
  "fold-before",
  "fold-immediate-history",
  "fold-history",
  "reopen-before",
  "reopen-history",
] as const;
const queryNames = [
  "process-restarted-before-fold",
  "fold-after-compaction",
  "reopen-after-compaction",
] as const;
const extraNames = [
  "seed",
  "fold-completion-pins",
  "fold-canonical-settlements",
  "reopen-canonical-settlements",
  "fold-contexts",
  "reopen-contexts",
  "fold-events",
  "create-process",
  "fold-process",
  "reopen-process",
  "create-result",
  "fold-result",
  "reopen-result",
] as const;

const sourceText =
  "TEST synthetic original testimony control: When final inspection starts, reserve one available crew until sign-off.";

interface JsonObject {
  [key: string]: unknown;
}

type RetentionBundle = Record<string, unknown>;

const isJsonObject = (value: unknown): value is JsonObject =>
  typeof value === "object" && value !== null && !Array.isArray(value);

const asObject = (value: unknown, label: string): JsonObject => {
  assert.ok(isJsonObject(value), `${label} must be an object`);
  return value;
};

const asArray = (value: unknown, label: string): unknown[] => {
  assert.ok(Array.isArray(value), `${label} must be an array`);
  return value;
};

const loadNamed = (directory: string, name: string): unknown => {
  const jsonPath = join(directory, `${name}.json`);
  if (existsSync(jsonPath)) {
    return JSON.parse(readFileSync(jsonPath, "utf8")) as unknown;
  }
  return JSON.parse(
    gunzipSync(readFileSync(join(directory, `${name}.json.gz`))).toString(
      "utf8",
    ),
  ) as unknown;
};

const textOf = (message: JsonObject): string =>
  asArray(message.parts, "message parts")
    .filter(
      (part): part is JsonObject =>
        isJsonObject(part) &&
        part.type === "text" &&
        typeof part.text === "string",
    )
    .map((part) => part.text)
    .join("");

const toolsOf = (snapshot: JsonObject): JsonObject[] =>
  asArray(snapshot.messages, "snapshot messages").flatMap((message) => {
    if (!isJsonObject(message)) {
      return [];
    }
    return asArray(message.parts, "message parts").filter(
      (part): part is JsonObject =>
        isJsonObject(part) && part.type === "dynamic-tool",
    );
  });

const toolOf = (snapshot: JsonObject, callId: string): JsonObject => {
  const found = toolsOf(snapshot).filter((part) => part.toolCallId === callId);
  assert.ok(
    found.length === 1 && found[0]?.state === "output-available",
    `exact completed tool: ${callId}`,
  );
  const tool = found[0];
  assert.ok(tool);
  return tool;
};

const require = (condition: unknown, message: string): void => {
  assert.ok(condition, message);
};

export const auditReopenedWhyRetention = (
  data: RetentionBundle,
): JsonObject => {
  const seed = asObject(data.seed, "seed");
  const pids = (["create", "fold", "reopen"] as const).map((phase) => {
    const processRecord = asObject(
      data[`${phase}-process`],
      `${phase} process`,
    );
    return processRecord.pid;
  });
  require(new Set(pids).size === 3, "distinct actual process IDs");
  for (const phase of ["create", "fold", "reopen"] as const) {
    const processRecord = asObject(
      data[`${phase}-process`],
      `${phase} process`,
    );
    const result = asObject(data[`${phase}-result`], `${phase} result`);
    require(result.pid === processRecord.pid, "process/result PID correlation");
    require(result.dbPath === seed.dbPath, "same original store path");
    require(result.outcome === "pass", "actual phase pass");
  }
  require(seed.pid === pids[0], "seed PID correlation");
  const baseline = asObject(data["create-history"], "create-history");
  const sourceMatches = asArray(baseline.messages, "baseline messages").filter(
    (message): message is JsonObject =>
      isJsonObject(message) && message.id === seed.sourceId,
  );
  require(sourceMatches.length === 1, "source exact identity/content");
  const source = sourceMatches[0];
  assert.ok(source);
  require(source.role === "user" &&
    source.purpose === "user", "source authorized role/purpose");
  require(textOf(source) === sourceText, "source exact identity/content");
  const protectedTools = toolsOf(baseline).filter((part) =>
    ["update_workpiece", "addArc", "getLatestNetDefinition"].includes(
      String(part.toolName),
    ),
  );
  require(protectedTools.length ===
    6, "three revisions, two reads, one mutation; no tool reissue");
  for (const number of [1, 2, 3]) {
    const revision = toolOf(baseline, `retention-revision-${number}`);
    const pointer = asObject(revision.output, "revision output");
    require(pointer.revisionId === revision.toolCallId &&
      pointer.ordinal === number, "revision identity/ordinal");
    require(isJsonObject(revision.input) &&
      typeof revision.input.markdown === "string" &&
      createHash("sha256").update(revision.input.markdown).digest("hex") ===
        pointer.sha256, "revision content hash");
    require(pointer.evidenceValidated === true, "validated revision evidence");
    const seedGoverning = asObject(seed.governing, "seed governing");
    require(JSON.stringify(pointer.evidence) ===
      JSON.stringify(seedGoverning.evidence) &&
      asArray(pointer.evidence, "revision evidence").length ===
        2, "overlapping carried evidence exact");
    if (number > 1) {
      require(isJsonObject(revision.input) &&
        !("evidence" in revision.input), "raw carried input not rewritten");
    }
  }
  const original = asObject(
    toolOf(baseline, "retention-live-why").output,
    "original why",
  );
  require(isJsonObject(original.reconciliation) &&
    original.reconciliation.status ===
      "live-observed", "creation actual live observation");
  for (const name of snapshotNames) {
    const snapshot = asObject(data[name], name);
    require(JSON.stringify(
      asArray(snapshot.messages, `${name} messages`).filter(
        (message) => isJsonObject(message) && message.id === seed.sourceId,
      ),
    ) === JSON.stringify([source]), "source exact identity/content");
    require(JSON.stringify(
      toolsOf(snapshot).filter((part) =>
        ["update_workpiece", "addArc", "getLatestNetDefinition"].includes(
          String(part.toolName),
        ),
      ),
    ) === JSON.stringify(protectedTools), "protected tools exact; no reissue");
    for (const message of asArray(baseline.messages, "baseline messages")) {
      if (!isJsonObject(message)) {
        continue;
      }
      require(JSON.stringify(
        asArray(snapshot.messages, `${name} messages`).filter(
          (entry) => isJsonObject(entry) && entry.id === message.id,
        ),
      ) === JSON.stringify([message]), "baseline public messages exact");
    }
  }
  for (const name of queryNames) {
    const query = asObject(data[name], name);
    const read = asObject(query.read, `${name} read`);
    require(JSON.stringify(read.currentWorkpiece) ===
      JSON.stringify(original.currentWorkpiece), "actual current state exact");
    for (const label of ["why", "oldObservationWhy"] as const) {
      const answer = asObject(query[label], `${name} ${label}`);
      require(JSON.stringify(answer.governing) ===
        JSON.stringify(
          original.governing,
        ), "governing revision/hash/passages/relations exact");
      require(JSON.stringify(answer.recordedChange) ===
        JSON.stringify(
          original.recordedChange,
        ), "actual recorded effects exact");
      require(isJsonObject(answer.reconciliation) &&
        answer.reconciliation.status ===
          "as-of", "restart is as-of, not fresh browser");
      require(answer.disposition === "partially-supported" &&
        answer.untrusted === true, "honest partial untrusted standing");
    }
    const oldWhy = asObject(query.oldObservationWhy, `${name} old why`);
    require(isJsonObject(oldWhy.reconciliation) &&
      oldWhy.reconciliation.observationScope ===
        "as-of", "old ID cannot earn freshness");
    require(asObject(query.refusedObservationWhy, `${name} refused`)
      .disposition === "refused", "unknown observation refuses");
    if (name !== "process-restarted-before-fold") {
      require(asArray(read.sources, `${name} sources`).some(
        (item) => isJsonObject(item) && item.id === seed.sourceId,
      ), "seed source remains discoverable");
      const phase = name.startsWith("fold") ? "fold" : "reopen";
      const request = asObject(
        asArray(data[`${phase}-contexts`], `${phase} contexts`)[
          Number(query.beforeRequestContextIndex)
        ],
        `${name} request context`,
      );
      require(request.purpose === "agent", "actual query request context");
      const context = asObject(request.context, `${name} context`);
      const serialized = JSON.stringify(context.messages);
      require(serialized.includes(
        "A5 controlled lossy summary",
      ), "real folded summary consumed");
      require(!serialized.includes(
        sourceText,
      ), "original true-user source entry absent from query context");
      require(!asArray(context.messages, `${name} context messages`).some(
        (message) =>
          isJsonObject(message) &&
          message.role === "toolResult" &&
          (message.toolName === "brunch_workpiece" ||
            message.toolName === "brunch_why"),
      ), "prior workpiece/why results absent from query context");
      require(!asArray(query.priorQueryIds, `${name} prior query ids`).some(
        (callId) => serialized.includes(String(callId)),
      ), "prior query IDs absent even with redacted source text");
    }
  }
  const foldEvents = asArray(data["fold-events"], "fold-events");
  const starts = foldEvents.filter(
    (event) => isJsonObject(event) && event.type === "compaction_start",
  );
  const compactions = foldEvents.filter(
    (event) => isJsonObject(event) && event.type === "compaction",
  );
  require(starts.length >= 2 &&
    starts.every(
      (event) => isJsonObject(event) && event.reason === "threshold",
    ), "real threshold compaction, never overflow substitute");
  require(compactions.length >= 2 &&
    compactions.every(
      (event) =>
        isJsonObject(event) &&
        event.isError === false &&
        Number(event.messagesAfter) < Number(event.messagesBefore),
    ), "successful real context folding");
  const pins = asArray(data["fold-completion-pins"], "fold-completion-pins");
  require(pins.length >= 22, "nonempty independent completion pins");
  for (const pinValue of pins) {
    const pin = asObject(pinValue, "completion pin");
    const records = asArray(pin.records, "pin records");
    const startRecords = records.filter(
      (record) =>
        isJsonObject(record) && record.type === "assistant_message_started",
    );
    const endRecords = records.filter(
      (record) =>
        isJsonObject(record) && record.type === "assistant_message_completed",
    );
    require(startRecords.length === 1 &&
      endRecords.length === 1, "canonical completion exists exactly once");
    const start = asObject(startRecords[0], "pin start");
    const end = asObject(endRecords[0], "pin end");
    const body = records
      .filter(
        (record): record is JsonObject =>
          isJsonObject(record) && record.type === "assistant_text_delta",
      )
      .map((record) => String(record.delta))
      .join("");
    const expected = {
      id: start.messageId,
      role: "assistant",
      purpose: "assistant",
      display: "visible",
      submissionId: start.submissionId,
      turnId: start.turnId,
      parts: [{ type: "text", text: body, state: "done" }],
    };
    require(JSON.stringify(pin.message) === JSON.stringify(expected) &&
      end.messageId === start.messageId &&
      end.stopReason ===
        "stop", "pin matches independent canonical completion");
    const event = asObject(pin.event, "pin event");
    require(event.turnId === start.turnId &&
      event.submissionId ===
        start.submissionId, "completion event correlation");
    for (const name of [
      "fold-history",
      "reopen-before",
      "reopen-history",
    ] as const) {
      const snapshot = asObject(data[name], name);
      require(JSON.stringify(
        asArray(snapshot.messages, `${name} messages`).filter(
          (message) => isJsonObject(message) && message.id === expected.id,
        ),
      ) ===
        JSON.stringify([
          expected,
        ]), "independently pinned completed response exact");
      require(JSON.stringify(
        asArray(snapshot.settlements, `${name} settlements`).filter(
          (item) =>
            isJsonObject(item) && item.submissionId === expected.submissionId,
        ),
      ) ===
        JSON.stringify([
          {
            submissionId: expected.submissionId,
            outcome: "completed",
            answeredBySubmissionId: expected.submissionId,
          },
        ]), "independently pinned completed settlement exact");
    }
    for (const phase of ["fold", "reopen"] as const) {
      const settlements = asArray(
        data[`${phase}-canonical-settlements`],
        `${phase} canonical settlements`,
      ).filter(
        (item) =>
          isJsonObject(item) && item.submissionId === expected.submissionId,
      );
      require(settlements.length === 1 &&
        isJsonObject(settlements[0]) &&
        settlements[0].outcome === "completed", "canonical settlement exact");
    }
  }
  return {
    pids,
    sameOriginalStore: seed.dbPath,
    completionPins: pins.length,
    thresholdCompactions: compactions.length,
    sourceId: seed.sourceId,
    governingRevision: asObject(seed.governing, "seed governing").revisionId,
  };
};

const falsifierCases = [
  ["source-omission", "source exact"],
  ["source-change", "source exact"],
  ["carried-relation-loss", "overlapping carried evidence"],
  ["completion-omission", "independently pinned completed response"],
  ["completion-duplicate", "independently pinned completed response"],
  ["settlement-omission", "independently pinned completed settlement"],
  ["same-pid", "distinct actual process"],
  ["false-live", "restart is as-of"],
  ["source-still-in-context", "original true-user source entry absent"],
  ["redacted-answer-still-in-context", "prior workpiece/why results absent"],
] as const;

export const falsifyReopenedWhyRetention = (
  original: RetentionBundle,
): {
  readonly mode: string;
  readonly rejected: true;
  readonly reason: string;
}[] => {
  const results: {
    readonly mode: string;
    readonly rejected: true;
    readonly reason: string;
  }[] = [];
  for (const [mode, expected] of falsifierCases) {
    const data = structuredClone(original);
    const seed = asObject(data.seed, "seed");
    const pins = asArray(data["fold-completion-pins"], "pins");
    const pin = asObject(asObject(pins[0], "first pin").message, "pin message");
    for (const name of snapshotNames) {
      const snapshot = asObject(data[name], name);
      if (mode === "source-omission") {
        snapshot.messages = asArray(
          snapshot.messages,
          `${name} messages`,
        ).filter(
          (message) => !(isJsonObject(message) && message.id === seed.sourceId),
        );
      }
      if (mode === "source-change") {
        for (const message of asArray(snapshot.messages, `${name} messages`)) {
          if (isJsonObject(message) && message.id === seed.sourceId) {
            message.parts = [
              { type: "text", text: "Mutated source", state: "done" },
            ];
          }
        }
      }
      if (mode === "carried-relation-loss") {
        const evidence = asArray(
          asObject(
            toolOf(snapshot, "retention-revision-2").output,
            "revision 2",
          ).evidence,
          "revision 2 evidence",
        );
        evidence.pop();
      }
      if (mode === "completion-omission") {
        snapshot.messages = asArray(
          snapshot.messages,
          `${name} messages`,
        ).filter(
          (message) => !(isJsonObject(message) && message.id === pin.id),
        );
      }
      if (
        mode === "completion-duplicate" &&
        asArray(snapshot.messages, `${name} messages`).some(
          (message) => isJsonObject(message) && message.id === pin.id,
        )
      ) {
        asArray(snapshot.messages, `${name} messages`).push(
          structuredClone(pin),
        );
      }
      if (mode === "settlement-omission") {
        snapshot.settlements = asArray(
          snapshot.settlements,
          `${name} settlements`,
        ).filter(
          (item) =>
            !(isJsonObject(item) && item.submissionId === pin.submissionId),
        );
      }
    }
    if (mode === "same-pid") {
      asObject(data["reopen-process"], "reopen process").pid = asObject(
        data["create-process"],
        "create process",
      ).pid;
    }
    if (mode === "false-live") {
      asObject(
        asObject(data["reopen-after-compaction"], "reopen query").why,
        "reopen why",
      ).reconciliation = {
        ...asObject(
          asObject(
            asObject(data["reopen-after-compaction"], "reopen query").why,
            "why",
          ).reconciliation,
          "reconciliation",
        ),
        status: "live-observed",
      };
    }
    if (mode === "source-still-in-context") {
      const query = asObject(
        data["reopen-after-compaction"],
        "reopen-after-compaction",
      );
      const contexts = asArray(data["reopen-contexts"], "reopen-contexts");
      const request = asObject(
        contexts[Number(query.beforeRequestContextIndex)],
        "reopen request",
      );
      const context = asObject(request.context, "reopen context");
      const baseline = asObject(original["create-history"], "create-history");
      const source = asArray(baseline.messages, "baseline messages").find(
        (message) => isJsonObject(message) && message.id === seed.sourceId,
      );
      assert.ok(isJsonObject(source));
      asArray(context.messages, "context messages").push({
        role: "user",
        content: textOf(source),
      });
    }
    if (mode === "redacted-answer-still-in-context") {
      const query = asObject(
        data["reopen-after-compaction"],
        "reopen-after-compaction",
      );
      const contexts = asArray(data["reopen-contexts"], "reopen-contexts");
      const request = asObject(
        contexts[Number(query.beforeRequestContextIndex)],
        "reopen request",
      );
      const context = asObject(request.context, "reopen context");
      const answer = structuredClone(asObject(query.why, "reopen why"));
      const governing = asObject(answer.governing, "governing");
      for (const passage of asArray(governing.passages, "passages")) {
        if (!isJsonObject(passage)) {
          continue;
        }
        for (const relation of asArray(passage.relations, "relations")) {
          if (isJsonObject(relation)) {
            relation.sources = [];
          }
        }
      }
      asArray(context.messages, "context messages").push({
        role: "toolResult",
        toolName: "brunch_why",
        toolCallId: "retention-live-why",
        content: [{ type: "text", text: JSON.stringify(answer) }],
      });
    }
    try {
      auditReopenedWhyRetention(data);
      throw new Error(`Falsifier escaped: ${mode}`);
    } catch (error) {
      assert.ok(error instanceof Error);
      require(error.message.includes(
        expected,
      ), `Wrong discriminator for ${mode}: ${error.message}`);
      results.push({ mode, rejected: true, reason: error.message });
    }
  }
  return results;
};

export const loadReopenedWhyRetention = (
  directory: string,
): RetentionBundle => {
  const names = [...snapshotNames, ...queryNames, ...extraNames];
  return Object.fromEntries(
    names.map((name) => [name, loadNamed(directory, name)]),
  );
};
